### DISCUSSION TITLE
Better usability & theming for SpinBox & the Tree SpinBox-like cell

### DISCUSSION THREAD

**Discussion by davthedev** [godotengine/godot-proposals#9260] [GUI]
I noticed some limitations with the existing implementation of SpinBox. Talking about this specific control with a text field and arrows to increase or decrease the value.

![image](https://github.com/godotengine/godot-proposals/assets/26961646/c1f55054-601b-4261-888a-d3bf8980f9e6)

The current limitations I found are:

- One graphic for the two "increase" and "decrease" actions although there are two distinct clickable areas
- No interaction feedback, we are not sure we hit the arrow correctly
- Limited style options, in comparison to what I want to achieve in a project

After some tinkering, I was able to get the following. The improvements added include

- Separate up and down buttons
- Interaction feedback when hovered and pressed
- When the value is at the extremity of the allowed range, one of the arrows is disabled
- Stylable arrow buttons & icons, separators, configurable minimum buttons width & spacing with the edit field
- The embedded LineEdit receives a theme variation name so that it is easy to style it a different way from the regular LineEdit using a global theme

Rendering on the default runtime theme

![spinbox_01](https://github.com/godotengine/godot-proposals/assets/26961646/2cd68fc8-b0e3-4562-8dec-a7f0c5b74e22)

With some styling, what is possible: (not the default theme, it is an example)

![spinbox_02](https://github.com/godotengine/godot-proposals/assets/26961646/3919bd51-219c-4936-ab56-35fc213d0554)

Big buttons for touch screens (customization example again)

![spinbox_03](https://github.com/godotengine/godot-proposals/assets/26961646/3b20a3c8-32c5-415a-abeb-27a29c7d6199)

Another possible minimal look with the new separator theme option

![spinbox_04](https://github.com/godotengine/godot-proposals/assets/26961646/7eadf8ee-a903-45ba-b354-b99c7d5463d8)

Live in Godot in the color selector

![spinbox_05](https://github.com/godotengine/godot-proposals/assets/26961646/86d44dd7-350f-4790-a034-532e0eee4d00)

Actually, there are two different widgets that render this UI element. The second one is a specialized cell of Tree, and seems to be used inside the inspector and project / editor options to set the various values.

Here are the improvement ideas.

- Interaction feedback && improvements above on the SpinBox control (PR available: https://github.com/godotengine/godot/pull/89265)
- Same on the Tree control for the SpinBox-like cell
- Styling the entire background of the widget encompassing the arrow buttons
- Allow setting the change speed, factor & delay when holding the mouse button pressed, same for the "value dragging" mode (press and hold on the buttons then move the mouse up/down) instead of hardcoded parameters
- Bring the Tree cell version in feature parity with the standalone SpinBox so that you can hold the mouse button pressed to give multiple "ticks". For now you have to click repeatedly. This would improve the properties editing usability a lot.
- Allow changing the side of the buttons independently of the RTL mode - would be "start" or "end"

**Comment by Zireael07** [comment]
Love the big buttons for the touchscreen (remember Godot itself is a Godot project AND has an Android version now)

**Reply by davthedev** [reply]
The big buttons are an example of what my changes in the PR can do. They are not brought to the default theme; you have to apply the style yourself.

**Comment by MewPurPur** [comment]
I know of multiple people who thought SpinBox was too scuffed to be useful, mainly because of this. So they winded up remaking this node from scratch, not even extending it. I'm one of them.

**Comment by akien-mga** [comment]
Implemented by https://github.com/godotengine/godot/pull/89265.
